src/library_pipefs.js

```diff
@@ -0,0 +1,208 @@
+          // Behave as if the read end is always non-blocking
+          throw new FS.ErrnoError(ERRNO_CODES.EAGAIN);
+        }
+        var toRead = Math.min(currentLength, length);
+
+        var totalRead = toRead;
+        var toRemove = 0;
+
+        for (var i = 0; i < pipe.buckets.length; i++) {
+          var currBucket = pipe.buckets[i];
+          var bucketSize = currBucket.offset - currBucket.roffset;
+
+          if (toRead <= bucketSize) {
+            var tmpSlice = currBucket.buffer.subarray(currBucket.roffset, currBucket.offset);
+            if (toRead < bucketSize) {
+              tmpSlice = tmpSlice.subarray(0, toRead);
+              currBucket.roffset += toRead;
+            } else {
+              toRemove++;
+            }
+            data.set(tmpSlice);
+            break;
+          } else {
+            var tmpSlice = currBucket.buffer.subarray(currBucket.roffset, currBucket.offset);
+            data.set(tmpSlice);
+            data = data.subarray(tmpSlice.byteLength);
+            toRead -= tmpSlice.byteLength;
+            toRemove++;
+          }
+        }
+
+        if (toRemove && toRemove == pipe.buckets.length) {
+          // Do not generate excessive garbage in use cases such as
+          // write several bytes, read everything, write several bytes, read everything...
+          toRemove--;
+          pipe.buckets[toRemove].offset = 0;
+          pipe.buckets[toRemove].roffset = 0;
+        }
+
+        while (toRemove--) {
+          pipe.buckets.shift();
+        }
+
+        return totalRead;
+      },
+      write: function (stream, buffer, offset, length, position /* ignored */) {
+        var pipe = stream.node.pipe;
+
+        assert(buffer instanceof ArrayBuffer || ArrayBuffer.isView(buffer));
+        var data = buffer.subarray(offset, offset + length);
+
+        var dataLen = data.byteLength;
+        if (dataLen <= 0) {
+          return 0;
+        }
+
+        var currBucket = null;
+
+        if (pipe.buckets.length == 0) {
+          currBucket = {
+            buffer: new Uint8Array(PIPEFS.BUCKET_BUFFER_SIZE),
+            offset: 0,
+            roffset: 0
+          };
+          pipe.buckets.push(currBucket);
+        } else {
+          currBucket = pipe.buckets[pipe.buckets.length - 1];
+        }
+
+        assert(currBucket.offset <= PIPEFS.BUCKET_BUFFER_SIZE);
+
+        var freeBytesInCurrBuffer = PIPEFS.BUCKET_BUFFER_SIZE - currBucket.offset;
+        if (freeBytesInCurrBuffer >= dataLen) {
+          currBucket.buffer.set(data, currBucket.offset);
+          currBucket.offset += dataLen;
+          return dataLen;
+        } else if (freeBytesInCurrBuffer > 0) {
+          currBucket.buffer.set(data.subarray(0, freeBytesInCurrBuffer), currBucket.offset);
+          currBucket.offset += freeBytesInCurrBuffer;
+          data = data.subarray(freeBytesInCurrBuffer, data.byteLength);
+        }
+
+        var numBuckets = ~~(data.byteLength / PIPEFS.BUCKET_BUFFER_SIZE);
```

Comment:
Interesting to see `~~` to truncate down to integer. Generally `x|0` is used throughout the codebase to achieve this, though I suppose these end up being functionally identical?